# **HOTEL BOOKING ANALYSIS**




##### **Project Type**    - EDA
##### **Contribution**    - Team
##### **Team Member 1 -**Dharmendra Yadav
##### **Team Member 2 -**Kratika Jawariya
##### **Team Member 3 -**Pranita tiwari


# **Project Summary**


The objective of this project is to conduct Exploratory Data Analysis (EDA) on a dataset related to hotel bookings.  By performing thorough exploratory analysis, we aim to uncover patterns, trends, and relationships within the dataset, which can provide valuable information to enhance decision-making processes in the hospitality industry.
The analysis focuses on several key aspects, including booking analysis, cancellation analysis, average daily rate (ADR) analysis, and customer retention.

#### **Data wrangling**:
The project begins with data wrangling tasks to ensure the data is in a suitable format for analysis. The data will be preprocessed to handle missing values, Removing duplicated columns, Created new columns and Changing data types. Relevant data cleaning techniques will be applied to ensure the quality of the dataset. Once the data wrangling phase is complete, the project moves on to the exploratory data analysis.

#### **Exploratory Data Analysis**:
The core phase of the project will involve conducting a comprehensive exploratory analysis. Throughout the analysis, visualizations using matplotlib and seaborn are employed to present the findings effectively. These visualizations aid in the interpretation and communication of the insights derived from the data. This will include investigating correlations between variables, identifying trends over time, and exploring patterns in customer behavior.

We will do the following type of analysis on our dataset:

**1. Customer Demographics**:

The analysis included factors such as customer nationality, market segments, and customer types (transient, contract, group). These insights can help tailor marketing strategies and target specific customer segments effectively.
Booking Patterns:

Understanding booking patterns, such as lead time, booking duration, and booking changes, provides insights into customer behavior and preferences. This information can guide revenue management strategies and optimize booking processes.

**2. Distribution Channels:**

The analysis examined different distribution channels, including online travel agents (OTA), direct bookings, and global distribution systems (GDS). Insights on revenue generation and booking cancellations from various channels can help hotels optimize their distribution strategies and improve revenue streams.

**3. Cancellations:**

The analysis delved into the reasons behind booking cancellations, exploring factors such as waiting time, lead time, and room type allocation. Identifying significant cancellation percentages for specific distribution channels (e.g., OTA) highlights areas for improvement and potential revenue loss.

**4. Revenue Generation:**

The analysis compared average daily rates (ADR) for different hotels, distribution channels, and months. Understanding revenue generation patterns can assist in pricing strategies, optimizing channel mix, and capitalizing on high-demand periods.

**5. Seasonal Trends:**

The analysis uncovered seasonal trends in terms of the number of guests, revenue, and ADR. Identifying peak months, such as August, can help hotels allocate resources effectively, manage staff levels, and tailor marketing campaigns to maximize occupancy and revenue during these periods.
The gained insights can positively impact the business in several ways:

**6. Data-Driven Decision-Making:**

By leveraging the insights from the analysis, hotels can make informed decisions based on data, enabling them to optimize pricing strategies, allocate resources effectively, and improve overall operational efficiency.
Revenue Management:

Understanding the factors influencing revenue generation, such as distribution channels and seasonal trends, empowers hotels to implement effective revenue management strategies. This includes adjusting pricing, managing inventory, and targeting specific customer segments to maximize revenue potential.

**7. Enhanced Customer Experience:**

The analysis of customer demographics and booking patterns provides valuable insights into customer preferences and behaviors. Hotels can use this information to tailor their services, improve guest experiences, and build customer loyalty.

**8. Marketing and Distribution Optimization:**

Insights on distribution channel performance, cancellation rates, and revenue generation can guide hotels in optimizing their marketing and distribution strategies. This may involve strengthening partnerships with high-performing channels, improving outreach on underperforming channels, and addressing cancellation issues through targeted measures.

**9. Competitive Advantage:**

Utilizing data-driven insights allows hotels to gain a competitive edge in the market. By understanding customer preferences, optimizing pricing, and aligning marketing efforts with seasonal trends, hotels can position themselves strategically to attract more guests and achieve higher revenue compared to their competitors.

While the analysis provided valuable insights, it is essential to continuously monitor and update the analysis based on changing market dynamics and evolving customer behaviors. Additionally, integrating external data sources, such as market trends and competitor analysis, can further enhance the depth and accuracy of the analysis.

The findings from the EDA will be summarized in a comprehensive report, supported by clear visualizations and actionable insights. The report will highlight key trends, patterns, and recommendations based on the data analysis. These insights can be leveraged by hotel management to optimize pricing strategies, improve customer satisfaction, and enhance overall operational efficiency.

# **Problem Statement**


<u>**Booking Analysis**</u>
- Which type of hotel has maximum booking?
- Which agent has done max number of bookings?
- Which Distibution channel lead to maximum booking?

<u>**Know about your hotel**</u>
- Which is most preferred meal?
- Which country do most guests come from??
- Which is the most preferred room type?
- How long people stay at hotel?

<u>**Revnue Analysis**</u>
- Which type of hotel seems to make more revnue?
- Which room type generates the hightes adr?
- Which distribution channel brings better revnue generating deals for hotel?

<u>**Cancellation analysis**</u>
- Which type of hotel has more cancellation?
- Which distribution channel has highest cancellation?
- Is there any impact of longer waiting days and lead time on cancellation?

<u>**Time wise Analysis**</u>
- To track month-on-month booking?
- To track the month-on-month adr of hotel?
- To track the year-on-year booking?

<u>**Customer retention**</u>
- How-many percentage of guest has arrived again?
- Which hotel has high chance that its customer will return for another stay?





#### **Define Your Business Objective?**

The objective of this project is to perfrom an EDA on hotel booking dataset to gain some insights which will be helpful in understanding the patterns and trends in booking analysis, cancellation analysis, Revnue analysis, customer segmentation and competitive analysis.
The valuable insights gained will be helpful in improving the operational efficiency, better finance management, increasing revnue, support data-driven decision-making and competitive positioning.

# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [ ]:
# import all the required python libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import datetime as dt

### Dataset Loading

In [ ]:
# mounting the google drive with the google colab
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# read csv file
path = '/content/drive/MyDrive/Dataset/Hotel Bookings.csv'
df = pd.read_csv(path)

### Dataset First View

In [ ]:
# first view of the dataset using head function and
df.head(10).T


0             1             2  \
hotel                           Resort Hotel  Resort Hotel  Resort Hotel   
is_canceled                                0             0             0   
lead_time                                342           737             7   
arrival_date_year                       2015          2015          2015   
arrival_date_month                      July          July          July   
arrival_date_week_number                  27            27            27   
arrival_date_day_of_month                  1             1             1   
stays_in_weekend_nights                    0             0             0   
stays_in_week_nights                       0             0             1   
adults                                     2             2             1   
children                                 0.0           0.0           0.0   
babies                                     0             0             0   
meal                                      BB            BB            BB   
country                                  PRT           PRT           GBR   
market_segment                        Direct        Direct        Direct   
distribution_channel                  Direct        Direct        Direct   
is_repeated_guest                          0             0             0   
previous_cancellations                     0             0             0   
previous_bookings_not_canceled             0             0             0   
reserved_room_type                         C             C             A   
assigned_room_type                         C             C             C   
booking_changes                            3             4             0   
deposit_type                      No Deposit    No Deposit    No Deposit   
agent                                    NaN           NaN           NaN   
company                                  NaN           NaN           NaN   
days_in_waiting_list                       0             0             0   
customer_type                      Transient     Transient     Transient   
adr                                      0.0           0.0          75.0   
required_car_parking_spaces                0             0             0   
total_of_special_requests                  0             0             0   
reservation_status                 Check-Out     Check-Out     Check-Out   
reservation_status_date           2015-07-01    2015-07-01    2015-07-02   

                                           3             4             5  \
hotel                           Resort Hotel  Resort Hotel  Resort Hotel   
is_canceled                                0             0             0   
lead_time                                 13            14            14   
arrival_date_year                       2015          2015          2015   
arrival_date_month                      July          July          July   
arrival_date_week_number                  27            27            27   
arrival_date_day_of_month                  1             1             1   
stays_in_weekend_nights                    0             0             0   
stays_in_week_nights                       1             2             2   
adults                                     1             2             2   
children                                 0.0           0.0           0.0   
babies                                     0             0             0   
meal                                      BB            BB            BB   
country                                  GBR           GBR           GBR   
market_segment                     Corporate     Online TA     Online TA   
distribution_channel               Corporate         TA/TO         TA/TO   
is_repeated_guest                          0             0             0   
previous_cancellations                     0             0             0   
previous_bookings_not_canceled             0             0             0   
reserved_room_

### **Dataset Rows & Columns count**





In [ ]:
#  Size of dataset
df.shape

(119390, 32)

### Dataset Information

In [ ]:
# check info i.e datatype and non of all the columns of dataset.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

What did you know about your dataset?

There are 119390 rows and 32 columns in this dataset.

## ***2. Understanding Your Variables***

In [ ]:
# Dataset Columns
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [ ]:
# Dataset Describe
df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000                   0.000000  
50%        94.57500

### Variables Description

1. **hotel**: Categorical - Resort hotel or city hotel.
2.	**is_canceled**: ‘1’ for booking and ‘0’  not cancelled.
3.	**lead_time**: Period between time of booking and checking in (considered in days here).
4.	**arrival_date_month**: Arrival month
5.	**country**: Country of origin. List of 158 countries.
6.	**days_in_waiting_list**: Number of waiting days.
7.	**Deposit_type**: Categorical - No-deposit, Non-Refund, Refundable.
8.	**Adr**: Average Daily rate as defined by the average rental revenue earned for an occupied room per day.
9.	**Adults, Babies, Children**: Number of adults, babies and children.
10.	**Assigned Room Type**: Code for the type of room assigned to the booking.
11.	**Booking Changes**: Number of changes/amendments made to the booking from the moment the booking was entered on the PMS until the moment of check-in or cancellation.
12.	**Distribution_channel**: Booking distribution channel.
13.	**Is_repeated_guest’**: Categorical- repeated guest(1) or not (0).
14.	**Company**: ID of the company/entity that made the booking or responsible for paying the booking.
15.	**Customer Type**: Categorical:
- Contract – when the booking has an allotment or other type of contract
associated to it
- Group – when the booking is associated to a group;
- Transient – when the bookings is not part of a group or contract, and is not associated to other transient booking;
- Transient party – when the booking is transient to at least other transient booking.
16.	**Market_segment**: Market segment designation.
17.	**Previous_cancellations**: Number of previous bookings that were cancelled by the customer prior to the current booking.
18.	**Required_car_parking_spaces**: Number of car parking spaces required by the customer.
19.	**Reservation_status**:Categorical:
- Cancelled – booking was cancelled by the customer.
- Check Out – customer has checked in but already departed.
- No Show – customer did not check in and did inform the hotel of the reason why.
20.	**Reservation_status_date**: Date at which the last status was set. This variable can be used in conjunction with the Reservation Status to understand when was the booking canceled or when did the customer checked-out of the hotel.
21.	**Reserved_room_type**: Code of room type reserved.
22.	**Types_of_special_requests**: Number of special requests made by the customer (e.g. Twin bed or high floor)
23.	**Stays_in_weekend_nights**, **Stays_in_week_nights**: Number of weekend nights and week nights the guest stayed or booked to stay at the hotel.


### Check Unique Values for each variable.

In [ ]:
# Checking the unique values of all the column using for loop
for i in df.columns:
  print(i, ':',df[i].unique(), "\n")

hotel : ['Resort Hotel' 'City Hotel'] 

is_canceled : [0 1] 

lead_time : [342 737   7  13  14   0   9  85  75  23  35  68  18  37  12  72 127  78
  48  60  77  99 118  95  96  69  45  40  15  36  43  70  16 107  47 113
  90  50  93  76   3   1  10   5  17  51  71  63  62 101   2  81 368 364
 324  79  21 109 102   4  98  92  26  73 115  86  52  29  30  33  32   8
 100  44  80  97  64  39  34  27  82  94 110 111  84  66 104  28 258 112
  65  67  55  88  54 292  83 105 280 394  24 103 366 249  22  91  11 108
 106  31  87  41 304 117  59  53  58 116  42 321  38  56  49 317   6  57
  19  25 315 123  46  89  61 312 299 130  74 298 119  20 286 136 129 124
 327 131 460 140 114 139 122 137 126 120 128 135 150 143 151 132 125 157
 147 138 156 164 346 159 160 161 333 381 149 154 297 163 314 155 323 340
 356 142 328 144 336 248 302 175 344 382 146 170 166 338 167 310 148 165
 172 171 145 121 178 305 173 152 354 347 158 185 349 183 352 177 200 192
 361 207 174 330 134 350 334 283 153 197 133 241 1

## **3. Data Wrangling**


1. Check for duplicated rows and remove if there is any duplicate rows.
2. Handling missing values.
3. Add new columns.
4. Converting data type.

### 1. Duplicate Values

In [ ]:
# Dataset Duplicate Value Count
df[df.duplicated()].count()

hotel                             31994
is_canceled                       31994
lead_time                         31994
arrival_date_year                 31994
arrival_date_month                31994
arrival_date_week_number          31994
arrival_date_day_of_month         31994
stays_in_weekend_nights           31994
stays_in_week_nights              31994
adults                            31994
children                          31994
babies                            31994
meal                              31994
country                           31958
market_segment                    31994
distribution_channel              31994
is_repeated_guest                 31994
previous_cancellations            31994
previous_bookings_not_canceled    31994
reserved_room_type                31994
assigned_room_type                31994
booking_changes                   31994
deposit_type                      31994
agent                             27847
company                            1538


In [ ]:
# Dropping duplicate values
df.drop_duplicates(inplace = True)

In [ ]:
# checking the size of data after removing the duplicate values
df.shape

(87396, 32)

### 2. Missing Values/Null Values

In [ ]:
# Missing Values/Null Values Count
df.isnull().sum().sort_values(ascending = False)

company                           82137
agent                             12193
country                             452
children                              4
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
hotel                                 0
previous_cancellations                0
days_in_waiting_list                  0
customer_type                         0
adr                                   0
required_car_parking_spaces           0
total_of_special_requests             0
reservation_status                    0
previous_bookings_not_canceled        0
is_repeated_guest                     0
is_canceled                           0
distribution_channel                  0
market_segment                        0
meal                                  0
babies                                0
adults                                0
stays_in_week_nights                  0


**Company** This column has most of its values as
null. Lets drop this column.

**Agent** This column has some of its values as null value, this can be because this tickets might not be booked through agents. Lets convert the null values with 0.

**Country** This column has very few values as null. Lets replace this null value with others.

**Children** This column has only 4 null values which is negligible. There are many 0 values in this column that means null value are those which are somehow missed to be added. Since this is very small we can replace them with mean value.

In [ ]:
# Company column dropped
df=df.drop(['company'],axis=1)

In [ ]:
# Agent column null value replace with zero
df[['agent']].fillna(0)

agent
0         0.0
1         0.0
2         0.0
3       304.0
4       240.0
...       ...
119385  394.0
119386    9.0
119387    9.0
119388   89.0
119389    9.0

[87396 rows x 1 columns]

In [ ]:
# Country column null value replace with others
df['country'].fillna('others', inplace = True)

In [ ]:
# Children value replaced with mean value of children
df['children'].fillna(df['children'].mean(), inplace = True)

### 3. Adding new columns

Total stay is calculated by calculating weekend night and week night stays.

In [ ]:
# A new column is creadted to know the total no. of booking days by adding weekend night and week nights
df['total_stay'] = df['stays_in_weekend_nights'] + df['stays_in_week_nights']

Total member is calculated by adding number of adults, children, babies.

In [ ]:
# Add no. of adults, children and babies to get total member stayed
df['total_member'] = df['adults'] + df['children'] + df['babies']

In [ ]:
df[df['total_stay'] == 0]
#  ['total_stay']

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
167     Resort Hotel            0        111               2015   
168     Resort Hotel            0          0               2015   
196     Resort Hotel            0          8               2015   
...              ...          ...        ...                ...   
115483    City Hotel            0         15               2017   
117701    City Hotel            0          0               2017   
118029    City Hotel            0          0               2017   
118631    City Hotel            0         78               2017   
118963    City Hotel            0          1               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
167                  July                        28   
168                  July                        28   
196                  July                        28   
...                   ...                       ...   
115483               July                        27   
117701             August                        32   
118029             August                        33   
118631             August                        34   
118963             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
167                             6                        0   
168                             6                        0   
196                             7                        0   
...                           ...                      ...   
115483                          6                        0   
117701                          8                        0   
118029                         14                        0   
118631                         23                        0   
118963                         27                        0   

        stays_in_week_nights  adults  ...  agent  days_in_waiting_list  \
0                          0       2  ...    NaN                     0   
1                          0       2  ...    NaN                     0   
167                        0       2  ...  240.0                     0   
168                        0       1  ...  250.0                     0   
196                        0       2  ...    NaN                     0   
...                      ...     ...  ...    ...                   ...   
115483                     0       1  ...   14.0                     0   
117701                     0       2  ...  262.0                     0   
118029                     0       2  ...    NaN                     0   
118631                     0       1  ...   98.0                     0   
118963                     0       2  ...    NaN                     0   

          customer_type  adr required_car_parking_spaces  \
0             Transient  0.0                           0   
1             Transient  0.0                           0   
167           Transient  0.0                           0   
168           Transient  0.0                           0   
196           Transient  0.0                           0   
...                 ...  ...                         ...   
115483  Transient-Party  0.0                           0   
117701        Transient  0.0                           0   
118029        Transient  0.0                           0   
118631  Transient-Party  0.0                           0   
118963        Transient  0.0                           0   

       total_of_special_requests  reservation_status  reservation_status_date  \
0                              0           Check-Out               2015-07-01   
1                              0

In [ ]:
df[df['total_member'] == 0]

hotel  is_canceled  lead_time  arrival_date_year  \
2224    Resort Hotel            0          1               2015   
2409    Resort Hotel            0          0               2015   
3181    Resort Hotel            0         36               2015   
3684    Resort Hotel            0        165               2015   
3708    Resort Hotel            0        165               2015   
...              ...          ...        ...                ...   
115029    City Hotel            0        107               2017   
115091    City Hotel            0          1               2017   
116251    City Hotel            0         44               2017   
116534    City Hotel            0          2               2017   
117087    City Hotel            0        170               2017   

       arrival_date_month  arrival_date_week_number  \
2224              October                        41   
2409              October                        42   
3181             November                        47   
3684             December                        53   
3708             December                        53   
...                   ...                       ...   
115029               June                        26   
115091               June                        26   
116251               July                        28   
116534               July                        28   
117087               July                        30   

        arrival_date_day_of_month  stays_in_weekend_nights  \
2224                            6                        0   
2409                           12                        0   
3181                           20                        1   
3684                           30                        1   
3708                           30                        2   
...                           ...                      ...   
115029                         27                        0   
115091                         30                        0   
116251                         15                        1   
116534                         15                        2   
117087                         27                        0   

        stays_in_week_nights  adults  ...  agent  days_in_waiting_list  \
2224                       3       0  ...    NaN                     0   
2409                       0       0  ...    NaN                     0   
3181                       2       0  ...   38.0                     0   
3684                       4       0  ...  308.0                   122   
3708                       4       0  ...  308.0                   122   
...                      ...     ...  ...    ...                   ...   
115029                     3       0  ...    7.0                     0   
115091                     1       0  ...    NaN                     0   
116251                     1       0  ...  425.0                     0   
116534                     5       0  ...    9.0                     0   
117087                     2       0  ...   52.0                     0   

          customer_type     adr required_car_parking_spaces  \
2224    Transient-Party    0.00                           0   
2409          Transient    0.00                           0   
3181    Transient-Party    0.00                           0   
3684    Transient-Party    0.00                           0   
3708    Transient-Party    0.00                           0   
...                 ...     ...                         ...   
115029        Transient  100.80                           0   
115091        Transient    0.00                           1   
116251        Transient   73.80                           0   
116534  Transient-Party   22.86                           0   
117087        Transient    0.00                           0   

       total_of_special_requests  reservation_status  reservation_status_date  \
2224                           0           Check-Out               2015-10-06

### 4. Converting data type

In [ ]:
df['agent'] = df['agent'].astype('str')

### What all manipulations have you done and insights you found?

1. **Duplicate values:** All the duplicated values are dropped.
2. **Missing values/Null Values:** There were 4 columns which has missing values. They are handled in different way in each column depending on the number of missing value and based on the nature of the missing data..
3. **Adding new columns:**
- total staying days in hotels: The code calculates the total staying days in hotels by summing the values from the 'stays_in_weekend_nights' and 'stays_in_week_nights' columns. It creates a new column named 'total_stay' to store these calculated values.
- Adding total_member: The code computes the total number of people (guests) by summing the values from the 'adults', 'children', and 'babies' columns. It creates a new column named 'total_people' to store the total count of people.

4. **Converting data types**: The code converts the data types of the agent columns from float to object using the astype() method. This ensures that this columns contain string datatype as the agent type is categorical values.

Insights:

Duplicated values are dropped. Missing values are handled differently in each case.
The newly created 'total_stay' column provides a consolidated view of the total duration of stays in hotels.
The 'total_member' column offers insights into the overall count of guests for each booking.
The converted data types allow for more accurate analysis involving the 'agent' columns.